File: 2024-08/05.py
# Problem: 2053. Kth Distinct String in an Array
# Link: https://leetcode.com/problems/kth-distinct-string-in-an-array/
# Runtime: 65 ms (Beats 81.26%)
# Memory: 16.82 MB (Beats 35.47%)
# Time Complexity: O(n), where n is the length of the array
# Space Complexity: O(n), for storing the frequency count of each string

from typing import List

class Solution:
    def kthDistinct(self, arr: List[str], k: int) -> str:
        # Dictionary to store the frequency of each string
        frequency = {}
        
        # Iterate through the array to count the occurrences of each string
        for string in arr:
            if string in frequency:
                frequency[string] += 1
            else:
                frequency[string] = 1
        
        # Iterate through the array to find the kth distinct string
        for string in arr:
            if frequency[string] == 1:
                k -= 1
                # If k reaches 0, return the current string as the kth distinct string
                if k == 0:
                    return string
        
        # If there are fewer than k distinct strings, return an empty string
        return ""

# Example 1
arr1 = ["d", "b", "c", "b", "c", "a"]
k1 = 2
print(Solution().kthDistinct(arr1, k1))  # Output: "a"

# Example 2
arr2 = ["aaa", "aa", "a"]
k2 = 1
print(Solution().kthDistinct(arr2, k2))  # Output: "aaa"

# Example 3
arr3 = ["a", "b", "a"]
k3 = 3
print(Solution().kthDistinct(arr3, k3))  # Output: ""

File: 2024-09/05.cpp
// Problem: 2028. Find Missing Observations
// Link: https://leetcode.com/problems/find-missing-observations/
// Beats 50.00% with 100 ms runtime, 69.12% with 116.40 MB memory usage

/*
Time Complexity:
1. Finding the total sum requires iterating over the `rolls` array: O(m), where m is the length of `rolls`.
2. The loop that distributes the missing values runs for n iterations.
3. Hence, the overall time complexity is O(m + n).

Space Complexity:
1. The primary space used is for storing the result array `ans`, which requires O(n) space.
2. The input array `rolls` does not require additional space since it's read-only.
3. Thus, the overall space complexity is O(n).
*/

#include <iostream>
#include <vector>
#include <numeric>   // for accumulate
#include <algorithm> // for min

using namespace std;

class Solution
{
public:
    vector<int> missingRolls(vector<int> &rolls, int mean, int n)
    {
        // Calculate the sum of all n + m rolls
        int totalSum = mean * (n + rolls.size());

        // Find the sum of the rolls that are known
        int currentSum = accumulate(rolls.begin(), rolls.end(), 0);

        // The sum we need to get from the missing rolls
        int missingSum = totalSum - currentSum;

        // Check if the missingSum is achievable within the constraints
        if (missingSum < n || missingSum > 6 * n)
            return {}; // Impossible case, return an empty vector

        vector<int> missingRolls(n);

        // Distribute the missing sum into the n missing rolls
        for (int i = 0; i < n; i++)
        {
            // Ensure each missing roll is between 1 and 6
            int value = min(6, missingSum - (n - i - 1));
            missingRolls[i] = value;
            missingSum -= value; // Update the remaining sum
        }

        return missingRolls;
    }
};

int main()
{
    Solution solution;

    // Test case 1
    vector<int> rolls1 = {3, 2, 4, 3};
    int mean1 = 4, n1 = 2;
    vector<int> result1 = solution.missingRolls(rolls1, mean1, n1);
    for (int r : result1)
        cout << r << " "; // Expected output: [6, 6]
    cout << endl;

    // Test case 2
    vector<int> rolls2 = {1, 5, 6};
    int mean2 = 3, n2 = 4;
    vector<int> result2 = solution.missingRolls(rolls2, mean2, n2);
    for (int r : result2)
        cout << r << " "; // Expected output: [2, 3, 2, 2]
    cout << endl;

    // Test case 3
    vector<int> rolls3 = {1, 2, 3, 4};
    int mean3 = 6, n3 = 4;
    vector<int> result3 = solution.missingRolls(rolls3, mean3, n3);
    for (int r : result3)
        cout << r << " "; // Expected output: [] (empty array)
    cout << endl;

    return 0;
}
